src/components/forms/TextInput/TextInput.tsx

```diff
@@ -50,21 +55,83 @@ export const TextInput = ({
       'usa-input--success': isSuccess,
       'usa-input--small': isSmall,
       'usa-input--medium': isMedium,
+      'usa-masked': mask,
     },
     className
   )
 
-  return (
+  const [inputValue, setInputValue] = useState('')
+  const input = (
     <input
       data-testid="textInput"
       className={classes}
       id={id}
       name={name}
       type={type}
       ref={inputRef}
+      maxLength={
+        inputProps.maxLength || !mask ? inputProps.maxLength : mask.length
+      }
+      onChange={mask ? handleChange : inputProps.onChange}
+      value={mask ? inputValue : inputProps.value}
       {...inputProps}
     />
   )
+
+  const [maskValue, setMaskValue] = useState(mask)
+  const [iValue, setIValue] = useState(<i></i>)
+  function handleChange(e: React.ChangeEvent<HTMLInputElement>): void {
+    const value = e.target.value
+    const maskData = charset || mask
+    if (undefined === maskData) return
+    const strippedValue = charset
+      ? value.replace(/\W/g, '')
+      : value.replace(/\D/g, '')
+    const isInteger = (v: string) => !Number.isNaN(parseInt(v, 10))
+    const isLetter = (v: string) => (v ? v.match(/[A-Z]/i) : false)
+    const maskedNumber = '_#dDmMyY9'
+    const maskedLetter = 'A'
+    let newValue = ''
+    for (let m = 0, v = 0; m < maskData.length; m++) {
+      const isInt = isInteger(strippedValue[v])
+      const isLet = isLetter(strippedValue[v])
```

Comment:
**nit**: high preference for spelling things out, but naming is nitpicky

```suggestion
      const charIsInteger = isInteger(strippedValue[v])
      const charIsLetter = isLetter(strippedValue[v])
```

Reply:
**suggestion (code flow)**: This also looks like you're trying to balance two separate conditions in parallel. 

I'm wondering if you can have `if (isInteger(...))` `else if (isLetter(...))` code blocks to separate the logic, rather than trying to keep track of two mutually exclusive things at the same time

Reply:
I cribbed the logic from the USWDS implementation, and it seems to simplify the _treatment_ at the expense of some complicated _conditions_. I attempted to simplify the conditions, but it didn't result in particularly better looking code though I didn't test performance.

Reply:
Ah, yea sometimes we crib from USWDS, that's fair. In those instances I'm happy to just let it be. Please do call that out for me and I'll forgo being nitpicky. We're a wrapper for their components after all. I believe in some place we have comments pointing to the source USWDS lines if we straight up copy paste their code, which is also fine by me.